tools/generator/dfg/stm32/stm_clock.py

```diff
@@ -0,0 +1,127 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2018, Niklas Hauser
+# All rights reserved.
+
+from os.path import commonprefix, basename
+import re
+import sys
+import logging
+import subprocess
+import tempfile
+import textwrap
+
+from collections import defaultdict
+from jinja2 import Environment
+from pathlib import Path
+from ..input.xml import XMLReader
+
+from . import stm
+
+LOGGER = logging.getLogger('dfg.stm32.clock')
+
+
+
+class STMClock:
+    ROOT_PATH = Path(__file__).parents[2]
+    CLOCK_FILE_PATH = ROOT_PATH / "raw-device-data/stm32-devices/plugins/clock"
+    CACHE_CLOCK_TREE = {}
+
+    @staticmethod
+    def get(did, rcc_ip_file):
+        if rcc_ip_file.filename not in STMClock.CACHE_CLOCK_TREE:
+            STMClock.CACHE_CLOCK_TREE[rcc_ip_file.filename] = STMClock(did, rcc_ip_file)
+        return STMClock.CACHE_CLOCK_TREE[rcc_ip_file.filename]
+
+    def __init__(self, did, rcc_ip_file):
+        self.ip_file = rcc_ip_file
+        self.did = did
+
+        match = basename(self.ip_file.filename)
+        match = re.search(r"RCC-STM32(((..).?.?)E?)[_-]rcc", match)
+        ip_name = match.group(1)
+        rcc_name = match.group(2)
+        family = match.group(3)
+        # print(family, rcc_name, ip_name)
+        files = [c for c in STMClock.CLOCK_FILE_PATH.glob("*.xml") if str(c).endswith("{}.xml".format(rcc_name))]
+        if not files:
+            files = [c for c in STMClock.CLOCK_FILE_PATH.glob("*.xml") if rcc_name in str(c)]
+        if not files:
+            files = [c for c in STMClock.CLOCK_FILE_PATH.glob("*.xml") if str(c).endswith("{}.xml".format(family))]
+        if len(files) != 1:
+            LOGGER.error("Unknown clock file for device '{}' and IP file '{}': {}".format(did.string, self.ip_file, files))
+            return
```

Comment:
Note to self: the ClockFile is encoded in the MCU XML at the *very top*: `<Mcu ClockTree="STM32F0"`
```xml
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<Mcu ClockTree="STM32F0" DBVersion="V3.0" Family="STM32F0" HasPowerPad="false" IOType="" Line="STM32F0x0 Value Line" Package="LQFP48" RefName="STM32F030C8Tx" xmlns="http://mcd.rou.st.com/modules.php?name=mcu">
	<Core>Arm Cortex-M0</Core>
	<Frequency>48</Frequency>
```